### ISSUE TITLE
Help: I am stuck trying to support Qtum on BTCPayServer

### ISSUE THREAD

**Issue by dabura667** [qtumproject/qtum#842]
I am trying to use QTUM with NBXplorer. (Backend of BTCPayServer)

NBXplorer keeps attempting to connect to my mainnet node and it is choking on some block headers between block 3999 and 5998

```
2020-06-18T12:56:39Z received: getheaders (773 bytes) peer=10
2020-06-18T12:56:39Z getheaders 3999 to end from peer=10
2020-06-18T12:56:39Z sending headers (434365 bytes) peer=10
```

It is just repeating these over and over every minute. So I think my NBitcoin (dep of NBX) implementation was flawed.

Does any of the consensus params here: https://github.com/MetacoSA/NBitcoin/blob/2330a5bf902bc44a78b779de88f14220954c6ab6/NBitcoin.Altcoins/Qtum.cs#L152-L191

Look suspicious?

Also,

I made the blockheader extended to include

32 byte hash,
32 byte hash,
36 byte outpoint (hash+vout)
and Script (push_data + signature)

Added on the default class: https://github.com/MetacoSA/NBitcoin/blob/2330a5bf902bc44a78b779de88f14220954c6ab6/NBitcoin.Altcoins/Qtum.cs#L43-L143

So I think this should work...

I think BTCPayServer support would be great... I think I am missing some parameter or something that maybe someone familiar with this can help.

**Comment by dabura667** [comment]
Maybe something happened around block 4000 that needs to be accounted for as a special case?

If anyone has an idea, please let me know. Thanks.

**Comment by dabura667** [comment]
I see 5001 block starts proof-of-stake... and from that point on the HASH256 of the header no longer is less than the target in the BITS section of the header... perhaps that is what is causing the issue.

**Comment by JB395** [comment]
Yes, I was going to suggest that. Qtum blocks 1 - 5000 are Proof of Work genesis blocks. Block 5001 starts Proof of Stake consensus, adding a coinstake transaction to each block, with the coinbase transaction empty.

**Comment by dabura667** [comment]
Yes. I checked the new transaction types, and as far as serializing and deserializing is concerned, they could be parsed by normal Transaction logic for Bitcoin transaction.

NBXplorer makes no consensus verifications and fully trusts the node it connects to, so it doesn't even check PoW for the block.

iirc it merely checks that GetHash of BlockHeader is equal to the prevblockhash of the next header.

I manually checked blocks before and after 5001 and they still use HASH256 (sha256 x 2)...

I would be lead to believe maybe it was choking on the coinstake tx if it wasn't choking on the initial getheaders loop. (the headers don't contain anything and all it does is hash and check that is equal next header's prevblock)

Perhaps the P2P message for `headers` message changed for blocks after 5000 and that is causing error? (maybe I need to change the P2P layer on NBX side?)

**Comment by qtum-neil** [comment]
There is no change in P2P messages. Please share full logs from the node (with debug=all), also logs from NBXplorer if any.

**Comment by dabura667** [comment]
I will try to get logs, for now I think I fixed the problem:

https://github.com/MetacoSA/NBitcoin/pull/874

**Comment by qtum-neil** [comment]
The fix looks good, the signature size is variable.

**Comment by dabura667** [comment]
I will let you know if this fixed the problem.

It was because I thought Script would deserialize as PUSH_DATA + SIG, but it actually was expecting VARINT_SCRIPTSIZE + PUSH_DATA + SIG, so it broke (I think)

So instead I am using the ReadWriteAsVarString which will read VARINT_LENGTH + LENGTH_BYTEARRAY instead.

So I think this is the issue. I will come back with logs if this doesn't work.

**Comment by dabura667** [comment]
Ok, this is a mixture of NBX logs and qtum logs.

Around this timestamp
2020-06-19T02:57:36Z
Loading Chain... is logged in NBX.

The you see NBX (peer=0) is doing getheaders loop, then you see it is retrying from 3999 over and over.

This was done with the fix included.

[output.txt](https://github.com/qtumproject/qtum/files/4802265/output.txt)

**Comment by qtum-neil** [comment]
Seems NBX is requesting the same headers over and over again, the node seems to send the requested headers correctly, so I would check on the NBX end why it does not receive/process the headers correctly, maybe by adding some debug code in the received headers processing code to see why it's not processed correctly.

**Comment by dabura667** [comment]
I found the issue:

The Outpoint class has a static constant Zero I was using.

I assumed it was (uint256)(0) and uint.MaxValue, but actually it was (uint)(0) even though the IsNull check checks against MaxValue.

I switched the default and it started working fine.

https://github.com/MetacoSA/NBitcoin/pull/875/files

Anywho, the current release of NBXplorer now seems to work for simple utxo based sends.
